src/datamodel/builder/UMessage.cpp

```diff
@@ -10,3 +10,224 @@
+                            std::move(source), std::move(method));
+
+    return builder.withPriority(priority).withTtl(ttl);
+}
+
+UMessageBuilder UMessageBuilder::response(v1::UUri&& sink, v1::UUID&& request_id,
+                                          v1::UPriority priority, v1::UUri&& method) {
+    auto [methodOk, methodReason] = UriValidator::isValid(method);
+    if (!methodOk) {
+        throw UriValidator::InvalidUUri(
+            "Method URI is not a valid URI |  " +
+            std::string(UriValidator::message(*methodReason)));
+    }
+    auto [sinkOk, sinkReason] = UriValidator::isValid(sink);
+    if (!sinkOk) {
+        throw UriValidator::InvalidUUri(
+            "Source URI is not a valid URI |  " +
+            std::string(UriValidator::message(*sinkReason)));
+    }
+    auto [reqIdOk, reqIdReason] = UUidValidator::isUuid(request_id);
+    if (!reqIdOk) {
+        throw UUidValidator::InvalidUuid(
+        "Request id is not a valid UUid | " +
+        std::string(UUidValidator::message(*reqIdReason)));
+    }
+
+    UMessageBuilder builder(v1::UMessageType::UMESSAGE_TYPE_RESPONSE,
+                            std::move(method), std::move(sink),
+                            std::move(request_id));
+
+    return builder.withPriority(priority);
+}
+
+UMessageBuilder UMessageBuilder::response(const v1::UMessage& request) {
+    v1::UUri sink = request.attributes().source();
+    v1::UUID reqId = request.attributes().reqid();
+    v1::UPriority priority = request.attributes().priority();
+    v1::UUri method = request.attributes().sink();
+
+    return UMessageBuilder::response(std::move(sink), std::move(reqId),
+                                      priority, std::move(method));
+
+}
+
+UMessageBuilder& UMessageBuilder::withPriority(v1::UPriority priority) {
+    if ((priority < v1::UPriority::UPRIORITY_CS0) ||
+        (priority > v1::UPriority::UPRIORITY_CS6)) {
+        throw std::out_of_range("Priority value is out of range");
+    }
+
+    if ((attributes_.type() == v1::UMessageType::UMESSAGE_TYPE_REQUEST) ||
+        (attributes_.type() == v1::UMessageType::UMESSAGE_TYPE_RESPONSE)) {
+        if (priority < v1::UPriority::UPRIORITY_CS4) {
+            throw std::out_of_range("Priority value is out of range, for req/resp should be > CS4");
+        }
+    }
+
+    attributes_.set_priority(priority);
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withTtl(std::chrono::milliseconds ttl) {
+    if ((ttl.count() <= 0) || (ttl.count() > std::numeric_limits<uint32_t>::max())) {
+        throw std::out_of_range("TTL value is out of range");
+    }
+
+    attributes_.set_ttl(static_cast<uint32_t>(ttl.count()));
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withToken(const std::string& token) {
+    if ((attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_REQUEST)) {
+        throw std::domain_error("Token can only be set on a request message");
```

Comment:
Was there a test for this throw ?  And the other domain_error throws ??  I looked, but might have missed it.

Reply:
Hehehe - I guess I jumped the gun a bit, I just now noticed that this PR is a draft -- I probably should have waited until you were ready for reviewers -- Whoops.

Reply:
Ohh. no worries, reviews are always welcome. In fact thank you for taking time. But yeah I am still working on unit tests.